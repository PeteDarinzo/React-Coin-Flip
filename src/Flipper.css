.Flipper {
  margin: 5%;
}

.Flipper-button {
  margin: 10px;
}