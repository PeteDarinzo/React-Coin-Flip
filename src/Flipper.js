import React, { useState } from "react";
import Coin from "./Coin";
import "./Flipper.css"

function Flipper(props) {

  /** hook for face-up side of coin */
  const [side, setSide] = useState(null);
  /** hook for coin flip statistics */
  const [count, setCount] = useState({
    total: 0,
    heads: 0,
    tails: 0
  });

  /** Determine face up side based on random 0 or 1 */
  const generateSide = () => {
    let side = ""
    let randBit = Math.floor(Math.random() * 2);
    if (randBit === 1) {
      side = "tails";
      count.tails++;
    } else {
      side = "heads";
      count.heads++;
    }
    return side;
  }

  /** Reset the coin and counters */
  const reset = () => {
    setSide(null);
    setCount({
      total: 0,
      heads: 0,
      tails: 0
    });
  }

  /** Handle coin flip upon button click
   * Explanation of using state with objects: https://react-hooks-cheatsheet.com/usestate
   */
  const flip = () => {
    // determine the new side, cast as a String to avoid errors when updating the count
    const side = String(setSide(generateSide()));
    // the entire object must be passed into the set function, since it replaced, not merged
    setCount({
      ...count, // spread all values into the new object, note that only the face-down side will be unchanged
      ["total"]: count["total"] + 1, // total count will always be updated
      [side]: count[side] + 1 // take the face up side, and increment
    });
  }

  return (
    <div className="Flipper">
      <h1>Flip a coin.</h1>
      <Coin face={side} />
      <button className="Flipper-button" onClick={flip}>Flip</button>
      <button className="Flipper-button" onClick={reset}>Reset</button>
      <p>Out of {count.total} flips, there have been {count.heads} heads and {count.tails} tails.</p>
    </div>
  );
}

export default Flipper;