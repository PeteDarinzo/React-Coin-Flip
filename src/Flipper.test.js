import { render, fireEvent, queryAllByText } from '@testing-library/react';
import Flipper from "./Flipper"

/**
 * Mock the Math.random function so that the first flip is always heads, and the second tails
 */
beforeEach(function () {
  jest
    .spyOn(Math, "random")
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75);
});

/**
 * Undo the mock after all tests
 */
afterEach(function () {
  Math.random.mockRestore();
});

it("renders without crashing", () => {
  render(<Flipper />)
});

it("matches the snapshot", () => {
  const { asFragment } = render(<Flipper />);
  expect(asFragment()).toMatchSnapshot();
});

it("does not initially show the coin", () => {
  const { getByAltText } = render(<Flipper />);
  expect(getByAltText("no coin")).toBeInTheDocument();
});

it("can flip the coin and update statistics", () => {
  const { queryByText, getByAltText } = render(<Flipper />);

  expect(queryByText("Out of 0 flips, there have been 0 heads and 0 tails.")).toBeInTheDocument();

  const flipButton = queryByText("Flip");

  // flip the coin once
  fireEvent.click(flipButton);

  expect(getByAltText("heads")).toBeInTheDocument();
  expect(queryByText("Out of 1 flips, there have been 1 heads and 0 tails.")).toBeInTheDocument();

  // flip again
  fireEvent.click(flipButton);

  expect(getByAltText("tails")).toBeInTheDocument();
  expect(queryByText("Out of 2 flips, there have been 1 heads and 1 tails.")).toBeInTheDocument();
});

it("can be reset", () => {
  const { queryByText, getByAltText } = render(<Flipper />);

  expect(queryByText("Out of 0 flips, there have been 0 heads and 0 tails.")).toBeInTheDocument();

  const flipButton = queryByText("Flip");

  // flip the coin twice
  fireEvent.click(flipButton);
  fireEvent.click(flipButton);

  expect(queryByText("Out of 2 flips, there have been 1 heads and 1 tails.")).toBeInTheDocument();

  const resetButton = queryByText("Reset");

  // reset the coin
  fireEvent.click(resetButton);

  expect(getByAltText("no coin")).toBeInTheDocument();
  expect(queryByText("Out of 0 flips, there have been 0 heads and 0 tails.")).toBeInTheDocument();
});