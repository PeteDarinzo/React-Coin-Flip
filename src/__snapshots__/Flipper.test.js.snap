// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="Flipper"
  >
    <h1>
      Flip a coin.
    </h1>
    <div
      class="Coin"
    >
      <img
        alt="no coin"
        class="Coin-image"
      />
    </div>
    <button
      class="Flipper-button"
    >
      Flip
    </button>
    <button
      class="Flipper-button"
    >
      Reset
    </button>
    <p>
      Out of 0 flips, there have been 0 heads and 0 tails.
    </p>
  </div>
</DocumentFragment>
`;
